runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java

```diff
@@ -214,24 +220,90 @@ public static Environment createProcessEnvironment(
       pathsToStage.addAll(stagingFiles);
     }
 
-    ImmutableList.Builder<ArtifactInformation> filesToStage = ImmutableList.builder();
+    ImmutableList.Builder<Supplier<ArtifactInformation>> lazyArtifactsBuilder =
+        ImmutableList.builder();
     for (String path : pathsToStage) {
```

Comment:
Don't we want this for loop to be lazy? 

Rather than introducing intermediate streams of Suppliers, I think we could just rename the existing `getArtifacts()` something like `getNonDeferredArtifacts()` and then call it during resolution. 

```
if (key.equals(deferredArtifactPayload.getKey())) {
      return getNonDeferredArtifacts(options);
}
```

Reply:
This for loop is fairly cheap and from stream of Suppliers we can easily get additional performance benefits by creating parallelStream. When we consider parallelizing expensive computations, some boilerplate codes are needed anyway in `getNonDeferredArtifacts()`. I think building a stream is a nice way to abstract them out.